### ISSUE TITLE
Detect change in KeyVault references

### ISSUE THREAD

**Issue by pratyushoak** [Azure/AppConfiguration-DotnetProvider#248]
I have an Azure Function, where we use Azure App Configuration to pull a certificate from our KeyVault. The AppConfiguration to KeyVault reference has been setup. The AppConfiguration registration looks something like this -
```
aadConfigBuilder.AddAzureAppConfiguration(options =>
            {
                options.Connect(Environment.GetEnvironmentVariable("AppConfiguration:ConnectionString"))
                    .ConfigureKeyVault(kv =>
                       {
                           kv.SetCredential(new DefaultAzureCredential());
                       })
                    .Select("ExternalSystemAuthentication:AadConfig:*")
                       .ConfigureRefresh(refreshOptions =>
                            refreshOptions
                            /* .Register("ExternalSystemAuthentication:AadConfig:AuthCert", refreshAll: true) */ // want to detect when this changes
                            .Register("ExternalSystemAuthentication:AadConfig:Sentinel", refreshAll: true) // this works as expected
                            .SetCacheExpiration(TimeSpan.FromMinutes(2)));
                configurationRefresher = options.GetRefresher();
            });
```
Everything works as expected when I use a value based 'Sentinel' key to detect a change in config. However, in my case I have auto-rotation setup on the KeyVault certificate. When a new version is available, I want the latest version to be pulled into my Function. However, since the KeyVault reference is just a URL its value does not change in the AppConfiguration, and this key cannot be used as the observable key. So I am forced to manually update the 'Sentinel' so that my function can refresh config. Is there a way App Configuration can detect a change in the KV secret value? Or is there a way I can force refresh of the App Configuration periodically without checking for a change in any key?
Thanks.

**Comment by avanigupta** [comment]
## Overview

Azure App Configuration Provider supports loading secrets from Key Vault into your application configuration based on the Key Vault references stored in your App Config store. Key Vault references are a special type of key-value in AppConfig, where the value is a URI that references the secret in Key Vault.

There are times when your application could be using secrets from Key Vault that are no longer valid. Here are some situations when your application should automatically reload secrets from Key Vault:
- A secret that the application is using has expired in Key Vault.
- A certificate that the application is using has an auto-renewal policy which caused the certificate to auto-rotate in Key Vault. 


## Problem

Azure App Configuration Provider does not automatically reload secrets in any of the above scenarios. There are only two situations where we load/reload secrets from KeyVault:
- Initial configuration loading during application start up.
- A sentinel key-value changed in App Config and it was registered for refresh with the `ConfigureRefresh` method (the sentinel key-value may or may not be a Key Vault reference).


## Proposed Solution

We can start caching secrets in application memory along with their expiry time (or auto-renewal time in case of certificates). 
Whenever application configuration is refreshed, we use the secret value from the cache if it has not expired. If it has expired, we can reload the secret from Key Vault and update the in-memory cache with the new secret value and expiry time.

To optimize the number of times Azure App Configuration Provider loads a secret from Key Vault, the following behavioral changes will be made:
- With every call to `RefreshAsync`, scan all the secrets present in the secret cache and reload those that have expired/rotated in Key Vault. If a secret has not expired, keep using the cached value, thereby reducing the number of requests made to Key Vault to fetch secrets.
- Do not reload the secret from Key Vault if a refresh registered key-value changes in AppConfig but `refreshAll` is `false`.

Additionally, if the cached secret expires and we try to reload the secret from KeyVault, we could hit one of these condtions:
- The secret/certificate was deleted from Key Vault.
- Application's permission to read the secret was revoked.

In these cases, the secret will be deleted from your secret cache and your application configuration.


## Application changes needed for auto-reloading certificates from Key Vault

Currently, we use a `SecretClient` to read both secrets and certificates from Key Vault. This approach works well for reading the value of a secret/certificate and the expiration time of a secret. However, this does not fetch the auto-renewal policy of certificates. For getting more details about when a certificate will be auto-rotated in Key Vault, we need to use a `CertificateClient`.

Therefore, if the Key Vault reference points to a certificate, we need an instance of `CertificateClient` which has the `Certificates - Get` permission to read certificates and their auto-renewal policy from Key Vault. Your application is responsible for providing the right credentials to read certificates from Key Vault. If these credentials are missing or do not have sufficient privileges, your certificates will not be auto-reloaded from Key Vault and no exception will be thrown.

There are two ways to provide credentials for fetching certificates from Key Vault:

### 1. EXISTING API: `AzureAppConfigurationKeyVaultOptions SetCredential(TokenCredential credential)`
If you are using this method to provide a default token credential to read secrets from Key Vault, the same credential can be used to read certificates from Key Vault too. Make sure that the credential you provide here has `Certificates - Get` permission in addition to `Secrets - Get` permission.

#### Usage example

```cs
var builder = new ConfigurationBuilder();
builder.AddAzureAppConfiguration(options =>
{
    options.Connect(connectionString)
           .Select("KeyVaultRef*")
           .ConfigureKeyVault(kv =>
           {
               // If you are using a default token credential for all Key Vault secrets, the same credentials will be used to read certificates too
               kv.SetCredential(new DefaultAzureCredential());
           });
```


### 2. NEW API: `AzureAppConfigurationKeyVaultOptions Register(CertificateClient certificateClient)`
We have introduced an overload for the existing `Register` method in `AzureAppConfigurationKeyVaultOptions` that accepts a `CertificateClient` instance. You can directly register all the `CertificateClient`s that your application might need for auto-reloading certificates from Key Vault. 

#### Usage example

```cs
var builder = new ConfigurationBuilder();
builder.AddAzureAppConfiguration(options =>
{
    options.Connect(connectionString)
           .Select("KeyVaultRef*")
           .ConfigureKeyVault(kv =>
           {
               // If you are registering a SecretClient to read Key Vault certificates, make sure to register a CertificateClient for reading their auto-renewal policy too
               kv.Register(new SecretClient(new Uri("https://abc.vault.azure.net"), new DefaultAzureCredential()));
               kv.Register(new CertificateClient(new Uri("https://abc.vault.azure.net"), new DefaultAzureCredential()));
           });
```


## Breaking Changes

This solution introduces the following changes that may break existing applications:
- Added an explicit dependency on `Azure.Security.KeyVault.Certificates` package version `4.1.0` .
- Changing a refresh registered key-value in App Config will not reload the secret from Key Vault unless `refreshAll` is `true` or unless the refresh registered key-value is a Key Vault reference.


cc: @drago-draganov @jimmyca15 @zhenlan

**Comment by avanigupta** [comment]
This feature has been released in package version 4.4.0:
- [Microsoft.Extensions.Configuration.AzureAppConfiguration](https://www.nuget.org/packages/Microsoft.Extensions.Configuration.AzureAppConfiguration/)
- [Microsoft.Azure.AppConfiguration.AspNetCore](https://www.nuget.org/packages/Microsoft.Azure.AppConfiguration.AspNetCore/)
